src/utils.jl

```diff
@@ -0,0 +1,43 @@
+function find_set!(disjoint_sets, i) # path compression
+    if i != disjoint_sets[i]
+        disjoint_sets[i] = find_set!(disjoint_sets, disjoint_sets[i])
+    end
+    return disjoint_sets[i]
+end
+
+function add_set!(disjoint_sets, weights_sets, i, j)
```

Comment:
What is `weights_sets` for?

Reply:
Is an optimizations for union-find to make the union and find operations almost constant.

Reply:
This is the fastest asymtotically variant known of union-find. The algorithm is described [here](https://algs4.cs.princeton.edu/15uf/).